#683 keychain sdk configure multiple nodes endpoints

backsapc opened the pull request:
https://github.com/warden-protocol/wardenprotocol/issues/683

backsapc committed:
Added support for multiple node urls

backsapc committed:
Changed config

backsapc committed:
Fixed test

pitasi commented on cmd/wardenkms/wardenkms.go:
```diff
@@ -38,6 +39,27 @@ type Config struct {
+type GrpcNodeConfigDecoder []GrpcNodeConfig
+
+func (sd *GrpcNodeConfigDecoder) Decode(value string) error {
+	smsProvider := make([]GrpcNodeConfig, 0)
```
not sure what smsProvider means :)

```suggestion
	nodesConfig := make([]GrpcNodeConfig, 0)
```

backsapc replied:
Fixed :)

pitasi commented on cmd/wardenkms/wardenkms.go:
```diff
@@ -38,6 +39,27 @@ type Config struct {
+
+type GrpcNodeConfigDecoder []GrpcNodeConfig
+
+func (sd *GrpcNodeConfigDecoder) Decode(value string) error {
```
let's avoid having a pointer to a pointer (= slice)

```suggestion
func (sd GrpcNodeConfigDecoder) Decode(value string) error {
```

backsapc replied:
Isn't it how the thing supposed to be?

pitasi commented on keychain-sdk/keychain.go:
```diff
@@ -99,30 +113,69 @@ func (a *App) Start(ctx context.Context) error {
 	}
 }
 
+func (a *App) liveTxClient() (*client.TxClient, error) {
```
what i'd like to see is a `Pool` type, that abstracts away from the caller the necessity of choosing a client

otherwise we leak the fact that we are connected to multiple nodes all the way down in the code

```go
type Pool struct {
  ...
}

func (p *Pool) BuildTx(...) { ... }

func (p *Pool) SendWaitTx(...) { ... }
```

this is all we need from a tx client

backsapc replied:
Fixed

pitasi commented on keychain-sdk/keychain.go:
```diff
@@ -99,30 +113,69 @@ func (a *App) Start(ctx context.Context) error {
+	initConnection := func(logger *slog.Logger, grpcNodeConfig GrpcNodeConfig, config BasicConfig, identity client.Identity) (*AppClient, error) {
+		appClient := &AppClient{
+			batchSize:    config.BatchSize,
+			keychainId:   config.KeychainID,
```
these are "global" config, they don't depend on the grpc endpoint, i'd rather not duplicate information

backsapc replied:
Fixed

pitasi commented on keychain-sdk/keychain.go:
```diff
@@ -99,30 +113,69 @@ func (a *App) Start(ctx context.Context) error {
-		return fmt.Errorf("failed to create query client: %w", err)
-	}
-	a.query = query
+	initConnection := func(logger *slog.Logger, grpcNodeConfig GrpcNodeConfig, config BasicConfig, identity client.Identity) (*AppClient, error) {
```
not a fan of this nested anonymous func, should we move it outside?

backsapc replied:
Fixed

pitasi commented on keychain-sdk/key_requests.go:
```diff
@@ -56,23 +56,30 @@ func (w *keyResponseWriter) Reject(reason string) error {
+		a.keyRequestTracker.Ingested(keyRequest.Id, client.grpcUrl)
+
+		if a.keyRequestTracker.HasReachedConsensus(keyRequest.Id, a.config.ConsensusNodeThreshold) {
+			keyRequestsCh <- keyRequest
```
isn't this called multiple times? 

e.g. if i have consensus threshold of 1 and am connected to 3 nodes, it will be called 3 times, by the 3 different goroutines? or am i missing something?

backsapc replied:
You're right, I fixed it. The problem is that if we keep removing requests from the tracker when they’re fulfilled, we risk executing them more than once due to possible stale nodes. So, we need to keep processed requests for a while, what you think?

backsapc replied:
@Pitasi never mind, @artur-abliazimov helped me realize it's not an issue, so I kept the deletion from the tracker as is.

pitasi reviewed (changes requested):
I think there are some changes to be made before merging this. Keychain SDK has been a brittle chunk of code in the past so we need to be careful and keep it as readable as possible.

I'd like to see some better abstractions over the fact that we are connected to multiple clients (i.e. a txClient pool), while the ingestion part could stay the same but it seems that the proposed version pushes duplicated requests to the channel.

backsapc commented on keychain-sdk/internal/tracker/tracker.go:
```diff
@@ -4,28 +4,53 @@ import (
+	}
+}
+
+func (t *T) IsNew(id uint64, nodeUrl string) bool {
```
Wrong

backsapc commented on cmd/wardenkms/wardenkms.go:
```diff
@@ -38,6 +38,27 @@
 
 	LogLevel slog.Level `env:"LOG_LEVEL, default=debug"`
+
+	ConsensusNodeThreshold uint `env:"CONSENSUS_NODE_THRESHOLD, default=1"`
```
It's okay I guess, that's not really critical at the moment

backsapc commented on cmd/wardenkms/wardenkms.go:
```diff
@@ -120,11 +149,14 @@
 	if cfg.HttpAddr != "" {
 		logger.Info("starting HTTP server", "addr", cfg.HttpAddr)
 		http.HandleFunc("/healthcheck", func(w http.ResponseWriter, r *http.Request) {
-			if app.ConnectionState() == connectivity.Ready {
-				w.WriteHeader(http.StatusOK)
-			} else {
-				w.WriteHeader(http.StatusServiceUnavailable)
+			for _, state := range app.ConnectionState() {
+				if state == connectivity.Ready {
+					w.WriteHeader(http.StatusOK)
+					return
+				}
 			}
+
+			w.WriteHeader(http.StatusServiceUnavailable)
```
@Pitasi what do you think?

artur-abliazimov commented on cmd/wardenkms/wardenkms.go:
```diff
@@ -56,18 +77,26 @@ func main() {
+			TxFees:        sdk.NewCoins(sdk.NewCoin("uward", math.NewInt(cfg.TxFee))),
+		},
+		GRPCConfigs:            grpcConfigs,
+		ConsensusNodeThreshold: cfg.ConsensusNodeThreshold,
```
`cfg.ConsensusNodeThreshold` is `uint`, but `ConsensusNodeThreshold` is `uint8`. Let's make them consistent.

backsapc replied:
Fixed

artur-abliazimov commented on cmd/wardenkms/wardenkms.go:
```diff
@@ -120,11 +149,14 @@ func main() {
-				w.WriteHeader(http.StatusOK)
-			} else {
-				w.WriteHeader(http.StatusServiceUnavailable)
+			for _, state := range app.ConnectionState() {
```
I think it would be more useful to return the state of each Grpc in the response body. Like,
```
[
  {
    "url": "url1",
    "state": "state1"
  },
  {
    "url": "url2",
    "state": "state2"
  }
]
```

artur-abliazimov commented on cmd/wardenkms/wardenkms.go:
```diff
@@ -144,3 +176,19 @@ func bigEndianBytesFromUint32(n uint64) ([4]byte, error) {
+		return nil, fmt.Errorf("GRPCUrls must be specified")
+	}
+
+	result := make([]keychain.GrpcNodeConfig, 0)
```
We know the length of the nodes array. Let's use it:
```
	var nodesLength = len(value)
	if nodesLength == 0 {
		return nil, fmt.Errorf("GRPCUrls must be specified")
	}

	result := make([]keychain.GrpcNodeConfig, 0, nodesLength)
```

backsapc replied:
Fixed

artur-abliazimov commented on cmd/wardenkms/wardenkms.go:
```diff
@@ -144,3 +176,19 @@ func bigEndianBytesFromUint32(n uint64) ([4]byte, error) {
+
+	result := make([]keychain.GrpcNodeConfig, 0)
+	for _, item := range value {
+		result = append(result, keychain.GrpcNodeConfig{
```
Let's ensure that `item.GRPCUrl` is not an empty string.

backsapc replied:
Fixed

artur-abliazimov commented on keychain-sdk/config.go:
```diff
@@ -58,3 +49,25 @@ type Config struct {
+	// gRPC server.
+	GRPCInsecure bool
+
+	// GRPCURLs is the list of URLs of the gRPC server to connect to.
```
GRPCURLs is the URL of the gRPC server to connect to.

backsapc replied:
fixed

artur-abliazimov commented on keychain-sdk/internal/tracker/status_tracker.go:
```diff
@@ -0,0 +1,53 @@
+
+const (
+	statusSeen      status = iota
+	statusProcessig        = iota
```
A type in `statusProcessig`. Must be `statusProcessing`.

backsapc replied:
fixed

backsapc committed:
Updated config for wardenkms

Lint fixes

Merge conflicts

Added ClientsPool abstraction

Build fixes

Lint fix

Lint fix

Update keychain-sdk/tx_client_pool.go

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>
Code review fixes

Merge branch 'feature/683-keychain-sdk-configure-multiple-nodes-endpoints' of github.com:warden-protocol/wardenprotocol into feature/683-keychain-sdk-configure-multiple-nodes-endpoints

Added removal of processed requests

Lint fix

backsapc committed:
Merge branch 'main' into feature/683-keychain-sdk-configure-multiple-nodes-endpoints

backsapc commented on keychain-sdk/example_keychain_test.go:
```diff
@@ -18,21 +18,24 @@ func Main() {
+		},
+
+		GRPCConfigs:            []keychain.GrpcNodeConfig{{GRPCURL: "localhost:9090", GRPCInsecure: false}},
+		ConsensusNodeThreshold: 1,
```
@Pitasi what do you think?

backsapc commented on keychain-sdk/internal/tracker/tracker.go:
```diff
@@ -4,28 +4,53 @@
 }
 
-func (t *T) Ingested(id uint64) {
+func (t *T) Ingested(id uint64, nodeUrl string) {
```
Fixed

backsapc commented on keychain-sdk/internal/tracker/tracker.go:
```diff
@@ -4,28 +4,53 @@
 )
 
+type hashSet map[string]struct{}
+
```
Fixed

backsapc commented on keychain-sdk/internal/tracker/tracker.go:
```diff
@@ -4,28 +4,53 @@
+		value = t.ingested[id]
+	}
+
+	value[nodeUrl] = struct{}{}
```
Won't fix

backsapc commented on keychain-sdk/keychain.go:
```diff
@@ -99,30 +114,69 @@
+			grpcInsecure: grpcNodeConfig.GRPCInsecure,
+		}
+
+		logger.Info("connecting to Warden Protocol using gRPC", "url", grpcNodeConfig, "insecure", grpcNodeConfig.GRPCInsecure)
```
Fixed

backsapc commented on cmd/wardenkms/wardenkms.go:
```diff
@@ -38,6 +38,27 @@ type Config struct {
+	GRPCInsecure bool
+}
+
+type GrpcNodeConfigDecoder []GrpcNodeConfig
```
Won't fix

backsapc commented on keychain-sdk/internal/tracker/status_tracker.go:
```diff
@@ -0,0 +1,62 @@
+		seensBy:   stringSet{},
+		status:    SEEN,
+	}
+}
```
Won't fix

backsapc commented on keychain-sdk/internal/tracker/status_tracker.go:
```diff
@@ -0,0 +1,62 @@
+	SEEN      status = iota
+	READY            = iota
+	PROCESSED        = iota
+)
```
Fixed

backsapc commented on keychain-sdk/internal/tracker/status_tracker.go:
```diff
@@ -0,0 +1,62 @@
+	sh.status = PROCESSED
+
+	return nil
+}
```
Method removed

backsapc commented on keychain-sdk/keychain.go:
```diff
@@ -64,23 +64,37 @@ func (a *App) SetSignRequestHandler(handler SignRequestHandler) {
 
+	a.clientsPool = clientsPool
+
+	a.txWriter = writer.New(
+		a.config.BatchSize,
+		a.config.BatchInterval,
+		a.config.TxTimeout,
+		a.logger())
+	a.txWriter.Fees = a.config.TxFees
+	a.txWriter.GasLimit = a.config.GasLimit
+
```
Won't fix

backsapc commented on keychain-sdk/keychain.go:
```diff
@@ -64,23 +64,37 @@
-	err := a.initConnections()
-	if err != nil {
+	clientsPool := NewClientsPool(a.config)
+	if err := clientsPool.initConnections(a.logger()); err != nil {
```
Won't fix

backsapc commented on keychain-sdk/keychain.go:
```diff
@@ -64,23 +64,37 @@
 	keyRequestsCh := make(chan *wardentypes.KeyRequest)
 	defer close(keyRequestsCh)
-	go a.ingestKeyRequests(keyRequestsCh)
+	for _, appClient := range a.clientsPool.clients {
+		go a.ingestKeyRequests(keyRequestsCh, appClient)
+	}
```
@Pitasi what do you think?

backsapc committed:
Fixed locking

backsapc committed:
Merge branch 'feature/683-keychain-sdk-configure-multiple-nodes-endpoints' of github.com:warden-protocol/wardenprotocol into feature/683-keychain-sdk-configure-multiple-nodes-endpoints

backsapc commented on keychain-sdk/internal/tracker/tracker.go:
```diff
@@ -4,32 +4,62 @@ import (
+		return ActionProcess, nil
+	}
+
+	return ActionSkip, nil
```
Fixed

backsapc commented on keychain-sdk/internal/tracker/tracker.go:
```diff
@@ -4,32 +4,62 @@
+		return t.ingested[id]
 	}
+
+	return value
```
Added Lock before calling this function

backsapc commented on keychain-sdk/keychain.go:
```diff
@@ -27,10 +27,20 @@ type App struct {
+	batchSize    int
+	keychainId   uint64
+	query        *client.QueryClient
+	txClient     *client.TxClient
```
Won't fix

backsapc commented on cmd/wardenkms/wardenkms.go:
```diff
@@ -38,6 +38,27 @@
+}
+
+type GrpcNodeConfig struct {
+	GRPCUrl      string
```
Won't fix

backsapc commented on keychain-sdk/internal/tracker/status_tracker.go:
```diff
@@ -0,0 +1,62 @@
+	}
+
+	return nil
+}
```
Won't fix

backsapc commented on keychain-sdk/internal/tracker/tracker.go:
```diff
@@ -4,32 +4,62 @@
+		return
+	}
+
+	_ = value.MarkProcessed()
```
Removed method

backsapc commented on keychain-sdk/internal/tracker/status_tracker.go:
```diff
@@ -0,0 +1,74 @@
+	threshold uint8
+	seensBy   stringSet
+	status    status
+}
```
Won't fix

backsapc commented on keychain-sdk/internal/tracker/tracker.go:
```diff
@@ -4,32 +4,57 @@ import (
+	}
+
+	return value
+}
```
Won't fix

backsapc committed:
Typo

backsapc commented on keychain-sdk/internal/tracker/tracker.go:
```diff
@@ -4,32 +4,57 @@
+		return ActionProcess, nil
+	}
+
+	return ActionSkip, nil
```
Fixed

backsapc commented on keychain-sdk/internal/tracker/status_tracker.go:
```diff
@@ -0,0 +1,53 @@
+const (
+	statusSeen      status = iota
+	statusProcessig        = iota
+)
```
Fixed

artur-abliazimov commented on keychain-sdk/internal/tracker/tracker.go:
```diff
@@ -4,32 +4,57 @@ import (
 	"sync"
 )
 
+type Action int
```
I think this solution can be simplified by deleting the `keychain-sdk/internal/tracker/status_tracker.go` and replace the `tracker`'s code with something like this:

```
package tracker

import (
	"sync"
)

type Action int

const (
	ActionSkip Action = iota
	ActionProcess
)

type T struct {
	threshold uint8
	rw        sync.RWMutex
	ingested  map[uint64]map[string]struct{}
}

func New(threshold uint8) *T {
	return &T{
		threshold: threshold,
		ingested:  make(map[uint64]map[string]struct{}),
	}
}

func (t *T) statusTracker(id uint64) map[string]struct{} {
	value, ok := t.ingested[id]
	if !ok {
		value = make(map[string]struct{})
		t.ingested[id] = value
	}
	return value
}

func (t *T) Ingest(id uint64, ingester string) (Action, error) {
	t.rw.RLock()
	defer t.rw.RUnlock()

	value := t.statusTracker(id)

	if _, exists := value[ingester]; exists {
		return ActionSkip, nil
	}

	value[ingester] = struct{}{}
	var ingesters = len(value)

	if ingesters == int(t.threshold) {
		return ActionProcess, nil
	}

	if ingesters >= int(t.threshold) {
		return ActionSkip, nil
	}

	return ActionSkip, nil
}

func (t *T) Done(id uint64) {
	t.rw.Lock()
	defer t.rw.Unlock()

	delete(t.ingested, id)
}
```

backsapc replied:
fixed

backsapc commented on keychain-sdk/internal/tracker/status_tracker.go:
```diff
@@ -0,0 +1,53 @@
+	status    status
+}
+
+type stringSet map[string]struct{}
```
Won't fix

backsapc commented on keychain-sdk/internal/tracker/status_tracker.go:
```diff
@@ -0,0 +1,53 @@
+		seensBy:   stringSet{},
+		status:    statusSeen,
+	}
+}
```
Won't fix

backsapc commented on keychain-sdk/keychain.go:
```diff
@@ -27,18 +26,19 @@ type App struct {
 		config:             config,
-		keyRequestTracker:  tracker.New(),
-		signRequestTracker: tracker.New(),
+		keyRequestTracker:  tracker.New(config.ConsensusNodeThreshold),
+		signRequestTracker: tracker.New(config.ConsensusNodeThreshold),
```
Won't fix

backsapc committed:
Simplified a bit

backsapc committed:
Code review

backsapc committed:
Code review

artur-abliazimov commented on keychain-sdk/tx_client_pool.go:
```diff
@@ -0,0 +1,124 @@
+	return liveClient.SendWaitTx(ctx, txBytes)
+}
+
+func (a *ClientsPool) ConnectionState() map[string]connectivity.State {
```
Let's rename all the mentions of the `(a *ClientsPool)` and `(p *ClientsPool)` to `(cp *ClientsPool)`.

backsapc replied:
Fixed

backsapc committed:
Changed status check

backsapc committed:
Fixed

backsapc committed:
Omit err check

backsapc committed:
Merge branch 'main' into feature/683-keychain-sdk-configure-multiple-nodes-endpoints

backsapc committed:
Merge branch 'main' into feature/683-keychain-sdk-configure-multiple-nodes-endpoints

backsapc committed:
Merge branch 'main' into feature/683-keychain-sdk-configure-multiple-nodes-endpoints

backsapc committed:
Merge branch 'main' into feature/683-keychain-sdk-configure-multiple-nodes-endpoints

pitasi commented on warden/x/act/client/actions.go:
```diff
@@ -141,7 +141,7 @@ func addFlagsFromMsg(msg sdk.Msg, cmd *cobra.Command) {
 			case reflect.Struct:
 				if v.Field(i).Type().Elem().AssignableTo(reflect.TypeOf(sdk.Coin{})) {
-					cmd.Flags().StringSlice(flagName, nil, "(e.g. 10uward)")
+					cmd.Flags().StringSlice(flagName, nil, "(e.g. 10award)")
```
this change is unrelated to this PR, please open a separate PR to fix bugs

pitasi commented on cmd/wardenkms/wardenkms.go:
```diff
@@ -56,18 +77,26 @@ func main() {
 		return
 	}
 
+	grpcConfigs, err := mapGrpcConfig(cfg.GRPCURLs)
```
since wardenkms is an user of the keychain-sdk, having to do this sort of mapping "manually" shows that the configuration of the keychain-sdk is not very ergonomic

i.e. keychain-sdk is not a library that I would enjoy using

i think we can explore different configuration patterns such as the option pattern maybe, or even treating the URL string as "special", like if it starts with "http://" or "tcp://" turn on the insecure option, while if it starts with "https://" turn insecure off

but, if you have better ideas please share them!

pitasi commented:
I don't know if I'm doing something wrong but I'm trying to run it with:
```
KEYRING_MNEMONIC="liberty ..." KEYRING_PASSWORD="$USER" GRPC_URLS='[{"GRPCUrl":"localhost:9090","GRPCInsecure":true}]' go run ./cmd/wardenkms
```

and it doesn't seem to be working? I tried logging the cfg:

```
logger.Info("XXXX", "grpc_urls", cfg.GRPCURLs)

time=2024-11-20T10:11:58.364+01:00 level=INFO msg=XXXX grpc_urls="[{GRPCUrl: GRPCInsecure:false} {GRPCUrl: GRPCInsecure:false}]"
```

and it's weird cause there are two "empty" elements (i.e. the url is not set)

how should I run it?

backsapc committed:
uward revert

backsapc committed:
Fixed wardenkms startup

backsapc commented:
> how should I run it?

You did everything right; the problem was on my end. I followed your suggestion and simplified the wardenkms configuration. I also fixed the startup issue.

pitasi commented on keychain-sdk/config.go:
```diff
@@ -7,23 +7,14 @@ import (
 
-// Config is the configuration for the Keychain.
-type Config struct {
+type BasicConfig struct {
```
what's the rationale behind splitting Config into Config and BasicConfig? I think I might be missing something

pitasi commented on keychain-sdk/config.go:
```diff
@@ -58,3 +49,25 @@ type Config struct {
+	// GRPCURL is the URL of the gRPC server to connect to.
+	// e.g. "localhost:9090"
+	GRPCURL string
+}
```
since we are inside a struct called "GrpcNodeConfig", let's avoid repeating the word GRPC everywhere:

```suggestion
type GrpcNodeConfig struct {
	// Insecure determines whether to allow an insecure connection to the
	// gRPC server.
	Insecure bool

	// Host is the address of the gRPC server to connect to.
	// e.g. "localhost:9090"
	Host string
}
```

technically, the second field is not a URL since it doesn't contain a protocol scheme, i renamed it into "host"

pitasi commented on keychain-sdk/config.go:
```diff
@@ -58,3 +49,25 @@ type Config struct {
+	GRPCConfigs []GrpcNodeConfig
+}
+
+type GrpcNodeConfig struct {
```
the case of the word grpc is inconsistent

```suggestion
type GRPCNodeConfig struct {
```

pitasi commented on keychain-sdk/example_keychain_test.go:
```diff
@@ -18,21 +18,24 @@ func Main() {
-		GRPCURL:      "localhost:9090",
-		GRPCInsecure: true,
+			// setup the connection to the Warden Protocol node
+			ChainID: "warden",
```
let's use this chance to update to our new default chain id:
```suggestion
			ChainID: "warden_1337-1",
```

pitasi commented on docker-compose.yml:
```diff
@@ -15,7 +15,7 @@ services:
       WARDEND_RPC_LADDR: tcp://0.0.0.0:26657
       WARDEND_RPC_CORS_ALLOWED_ORIGINS: "*"
-      WARDEND_MINIMUM_GAS_PRICES: "0uward"
+      WARDEND_MINIMUM_GAS_PRICES: "0award"
```
let's fix this in another PR since it's not really related to the keychain sdk

pitasi commented on keychain-sdk/config.go:
```diff
@@ -58,3 +49,25 @@ type Config struct {
+
+	// GRPCURLs is the list of URLs of the gRPC server to connect to.
+	// e.g. "localhost:9090"
+	GRPCConfigs []GrpcNodeConfig
```
since we call them "nodes" (Consensus**Node**Threshold represents the number of **nodes**...) we should be consistent. Also, the doc comment for GRPCConfigs is outdated.

```suggestion
	// ConsensusNodeThreshold represents the number of nodes required to execute a pending key/sign request.
	ConsensusNodeThreshold uint8

	// Nodes is the list of gRPC nodes to connect to for fetching incoming requests and broadcasting transactions.
	Nodes []GrpcNodeConfig
```

pitasi commented on cmd/wardenkms/health_check_response.go:
```diff
@@ -0,0 +1,23 @@
+package main
+
+type HealthCheckResponse struct {
+	// The number of nodes that are online
+	Online uint `json:"online"`
+
+	// The total number of nodes
+	Total uint `json:"total"`
+
+	// The consensus threshold
+	Threshold uint8 `json:"threshold"`
+
+	// Node statuses
+	Nodes []NodeStatus `json:"nodes"`
+}
+
+type NodeStatus struct {
+	// The address of the node
+	Address string `json:"address"`
+
+	// The status of the node
+	Status string `json:"status"`
+}
```
Let's follow Go conventions when documenting types or fields: https://go.dev/wiki/CodeReviewComments#comment-sentences.

```suggestion
type HealthCheckResponse struct {
	// Online is...
	Online uint `json:"online"`

	// Total is...
	Total uint `json:"total"`

	// Threshold is...
	Threshold uint8 `json:"threshold"`

	// Nodes...
	Nodes []NodeStatus `json:"nodes"`
}

type NodeStatus struct {
	// Address...
	Address string `json:"address"`

	// Status...
	Status string `json:"status"`
}
```

pitasi commented on keychain-sdk/tx_client_pool.go:
```diff
@@ -0,0 +1,124 @@
+	msgers ...client.Msger) ([]byte, error) {
+	liveClient, err := p.liveTxClient()
+	if err != nil {
+		return nil, fmt.Errorf("failed to aquire live client for BuildTx: %w", err)
```
Fix typo here

pitasi commented on keychain-sdk/tx_client_pool.go:
```diff
@@ -0,0 +1,124 @@
+func (cp *ClientsPool) SendWaitTx(ctx context.Context, txBytes []byte) error {
+	liveClient, err := cp.liveTxClient()
+	if err != nil {
+		return fmt.Errorf("failed to aquire live client for SendWaitTx: %w", err)
```
```suggestion
		return fmt.Errorf("failed to acquire live client for SendWaitTx: %w", err)
```

pitasi commented on keychain-sdk/tx_client_pool.go:
```diff
@@ -0,0 +1,124 @@
+	"google.golang.org/grpc/connectivity"
+)
+
+type AppClient struct {
```
Let’s avoid using the term “app” which is overloaded (in this case we’re connecting to a warden node, it can be confused with the keychain app itself), plus avoid exporting it because as far as I understand the use won’t initialize this manually

```suggestion
type wardenClient struct {
```

backsapc committed:
Update keychain-sdk/tx_client_pool.go

Co-authored-by: Antonio Pitasi <antonio@pitasi.dev>

backsapc committed:
Update keychain-sdk/example_keychain_test.go

Co-authored-by: Antonio Pitasi <antonio@pitasi.dev>

backsapc committed:
Merge branch 'main' into feature/683-keychain-sdk-configure-multiple-nodes-endpoints

backsapc committed:
Code review comments

backsapc committed:
Merge branch 'main' into feature/683-keychain-sdk-configure-multiple-nodes-endpoints

backsapc committed:
Merge remote-tracking branch 'origin/main' into feature/683-keychain-sdk-configure-multiple-nodes-endpoints

backsapc committed:
Updates

backsapc committed:
Fixed errors

backsapc committed:
Formatting

backsapc committed:
Code format fixes

backsapc commented on keychain-sdk/keychain.go:
```diff
@@ -65,23 +65,37 @@ func (a *App) SetSignRequestHandler(handler SignRequestHandler) {
 	keyRequestsCh := make(chan *wardentypes.KeyRequest)
 	defer close(keyRequestsCh)
-	go a.ingestKeyRequests(ctx, keyRequestsCh)
+	for _, appClient := range a.clientsPool.clients {
+		go a.ingestKeyRequests(ctx, keyRequestsCh, appClient)
+	}
 
 	signRequestsCh := make(chan *wardentypes.SignRequest)
 	defer close(signRequestsCh)
-	go a.ingestSignRequests(ctx, signRequestsCh)
+	for _, appClient := range a.clientsPool.clients {
+		go a.ingestSignRequests(ctx, signRequestsCh, appClient)
+	}
 
```
The code runs for infinite amount of time until the `context.Done()`. The issue you're highlighting here won't happen.

pitasi commented on keychain-sdk/internal/tracker/tracker.go:
```diff
@@ -1,35 +1,72 @@
+	value := t.ingestTracker(id)
+
+	if !value.add(ingesterId) {
+		return ActionSkip, errors.New("already ingested")
```
this is not an error, as it is expected to ingest the same id multiple times from the same source

pitasi commented on keychain-sdk/internal/writer/writer.go:
```diff
@@ -36,23 +33,26 @@ type W struct {
 }
 
-func (w *W) Start(ctx context.Context, flushErrors chan error) error {
+func (w *W) Start(ctx context.Context, client SyncTxClient, flushErrors chan error) error {
```
why removing the `Client` field from `W` to add it back as a parameter here (and in the other methods)?

pitasi commented on keychain-sdk/key_requests.go:
```diff
@@ -97,6 +113,6 @@ func (a *App) handleKeyRequest(ctx context.Context, keyRequest *wardentypes.KeyR
 
-func (a *App) keyRequests(ctx context.Context) ([]*wardentypes.KeyRequest, error) {
-	return a.query.PendingKeyRequests(ctx, &client.PageRequest{Limit: uint64(a.config.BatchSize)}, a.config.KeychainID)
+func (a *wardenClient) keyRequests(ctx context.Context, batchSize int, keychainId uint64) ([]*wardentypes.KeyRequest, error) {
+	return a.query.PendingKeyRequests(ctx, &client.PageRequest{Limit: uint64(batchSize)}, keychainId)
 }
```
if this is a method of `wardenClient`, let's move it near its definition (in `tx_client_pool.go`)

pitasi commented on keychain-sdk/sign_requests.go:
```diff
@@ -118,6 +129,6 @@ func (a *App) handleSignRequest(ctx context.Context, signRequest *wardentypes.Si
 
-func (a *App) signRequests(ctx context.Context) ([]*wardentypes.SignRequest, error) {
-	return a.query.PendingSignRequests(ctx, &client.PageRequest{Limit: uint64(a.config.BatchSize)}, a.config.KeychainID)
+func (a *wardenClient) signRequests(ctx context.Context, batchSize int, keychainId uint64) ([]*wardentypes.SignRequest, error) {
+	return a.query.PendingSignRequests(ctx, &client.PageRequest{Limit: uint64(batchSize)}, keychainId)
 }
```
if this is a method of `wardenClient`, let's move it near its definition (in `tx_client_pool.go`)

pitasi commented on keychain-sdk/internal/tracker/tracker.go:
```diff
@@ -1,35 +1,72 @@
-	rw       sync.RWMutex
-	ingested map[uint64]struct{}
+	threshold uint8
+	rw        sync.RWMutex
```
looks like we no longer need to use a `RWMutex`, a simple `Mutex` will do

pitasi commented on keychain-sdk/internal/tracker/tracker.go:
```diff
@@ -1,35 +1,72 @@
+
+type stringSet map[string]struct{}
+
+// add safely adds a string to the set.
```
there's no "safety" here, i suggest we rephrase it like this:

```suggestion
// add adds a string to the set and returns true if the key was already present.
```

pitasi commented on keychain-sdk/key_requests.go:
```diff
@@ -46,30 +47,45 @@ func (w *keyResponseWriter) Reject(ctx context.Context, reason string) error {
+	}
+
+	if action == tracker.ActionProcess {
+		keyRequestsCh <- keyRequest
```
duplicate keyRequests should NOT be written to keyRequestsCh, but this is not being guaranteed:
- N goroutines are running ingestRequest() concurrently
- threshold will most likely be set to T < N, e.g. T = 1
- all the N goroutines at some point will Ingest the same keyRequest, and all will write the same request to keyRequestsCh

pitasi reviewed (commented):
leaving some comments, mostly i'm concerned about the duplicate requests being written to the channels (both key requests and sign requests)

pitasi reviewed (changes requested):
leaving some comments, mostly i'm concerned about the duplicate requests being written to the channels (both key requests and sign requests)
